src/comm.cc

```diff
@@ -468,7 +468,7 @@ comm_apply_flags(int new_socket,
     if ( (flags & COMM_DOBIND) || addr.port() > 0 || !addr.isAnyAddr() ) {
         if ( !(flags & COMM_DOBIND) && addr.isAnyAddr() )
             debugs(5, DBG_IMPORTANT,"WARNING: Squid is attempting to bind() port " << addr << " without being a listener.");
-        if ( addr.isNoAddr() )
+        if (!addr.isBindable())
```

Comment:
This change caused a problem: the new condition evaluates to 'true' for unspecified (i.e., 'any') address and the debug message below is displayed. This logic seems correct: we report about 'not-bindable' address. The isBindable(), which is equivalent to !isEmpty() && !isNoAddr() seems correct either. There problem here, IMO is that Ip::Address cannot distinguish two states: isEmpty() and isAnyAddr(). For example, if we create an empty Ip::Address and call setAnyAddr() on it, isEmtpy() still returns true. I suspect that the roots of this problem come from the fact that IN_ADDR_ANY is usually 0 and leaving zero sin_addr/sin6_addr when initializing sockaddr_in/sockaddr_in6 structures also means 'bind to any address'. So, passing a brand new Ip::Address to bind() has the same effect as calling setAnyAddr() and bind(). I wonder whether we need some new flag in Ip::Address  to distinguish these two states.

Reply:
If isAnyAddr() usually implies isEmpty() and isEmpty() usually implies isAnyAddr(), but the calling code assumes that those are always two different states, than something went wrong. Moreover, isAnyAddr() documentation already says (badly) that "any address" is the initial state of an Ip::Address object. We should not be adding isEmpty() which claims to test the same thing.

You have added isEmpty() for some specific reason. What was that reason? I did not find that reason in commit 1d9260b. And why do we need isEmpty() today?

BTW, commit 1d9260b appears to be changing the meaning of the HttpRequest::indirectClientAddr() code because `IpAddress::operator==` used by isEmpty() ignores port differences, and the caller looked at the port before that commit. I did not trace what happened to that calling code; perhaps it is already fixed. It was the only isEmpty() user at the commit 1d9260b time.

I do not know whether it is important or not right now, but it sounds like isBindable() simply means !noAddr() -- we can call bind() with all other addresses AFAICT.

Commit a3315b8 talks about distinguishing the three special addresses, but it appears to be based on the false assumption that there are three special addresses (instead of two), and it is difficult for me to extract useful information (such as meaningful examples that would justify isEmtpy() existence) from that commit message.

> I wonder whether we need some new flag in Ip::Address to distinguish these two states.

Before we discuss the (very remote!) possibility of adding flags to Ip::Address, let's clarify: Why do we need (or "What code needs...") to distinguish the following two Ip::Address states?

1. an Ip::Address carrying `IN_ADDR_ANY` IP address and any port value
2. default-created Ip::Address

If that is not the distinction we are after, then please state what problem we are trying to solve by adding isEmpty().

Reply:
I think we need to consider Ip::Address use cases, before deciding whether isEmpty() (and isAnyAddr()) is necessary/applicable/different_meaning etc.  In general, I assume that 'any addr' (and underlying INADDR_ANY socket option) may be specified for server (bindable) socket only.

1. Local (listening server) address, passed to bind(). isEmpty() is not needed here and isAnyAddr() works well instead.
   isBindable() now returns just "!isNoAddr()" thus fixing the underlying problem.

2. Local(client) addresses used in self-generated requests (Downloader, etc). isAnyAddr() has no sense for clients: we probably need to call isEmtpy() (though the both methods returning true correspond to the same Ip::Adress state). Note that before the branch changes such addresses were configured as 'no addr'.

3. Remote client/server addresses. Same as (2): if these addresses cannot be acquired, I think we need to use isEmpty() API, emphasizing that the Ip::Address is still in its pristine state. Before the branch changes some (or may be all) contexts used 'no addr' if the remote address was unknown (for example, when responding with error page).

So, in short, I think that isAnyAddr() is applicable to the local(server) addresses, while isEmtpy() - to all other addresses. If I am right, we probably need some indication in Ip::Address about this kind, so that we were obliged to call a correct method in each context. For now, we can leave just a comment about that.

Reply:
As we discussed, it would be better to eliminate isEmpty() method. At the same time, we still need an indication (other from the inapplicable isAnyAddr()) of whether client address 'is empty', i.e., still has a default value. After some reflection, I dismissed the approach we discussed previously (due to implementation complexity) of creating a 'proxy' class for Ip::Address, which would keep the information about such 'emptiness'.  Instead, I used the isKnown() method in the two affected contexts (37bd940b). This method is used already in many other contexts in this branch (and probably should have been used in these two contexts also).
Basically, we have two 'higher level' Address validity checking methods: isBindable() and isKnown(). The first one should be used for checking (1) [type](https://github.com/measurement-factory/squid/pull/20/files#r354874746) while the second - for (2) and ( 3) types.